dali/pipeline/operator/eager_operator.h

```diff
@@ -35,42 +36,87 @@
 namespace dali {
 
 template <typename Backend>
-std::shared_ptr<TensorList<Backend>> AsTensorList(const std::shared_ptr<TensorList<Backend>> &in) {
+std::shared_ptr<TensorList<Backend>> AsTensorList(const std::shared_ptr<TensorList<Backend>> in) {
   return in;
 }
 
 template <typename Backend>
-std::shared_ptr<TensorList<Backend>> AsTensorList(
-    const std::shared_ptr<TensorVector<Backend>> &in) {
+std::shared_ptr<TensorList<Backend>> AsTensorList(const std::shared_ptr<TensorVector<Backend>> in) {
   if (in->IsContiguous()) {
     // Filled contiguous TensorVector, we can return TensorList directly.
-    return in->AsTensorList(false);
+    auto tl = in->AsTensorList(false);
+    // Explicitly set layout (it could be empty in case of per-sample operators).
+    tl->SetLayout(in->GetLayout());
+    return tl;
   }
 
   auto tl = std::make_shared<TensorList<Backend>>();
   tl->Copy(*in);
   return tl;
 }
 
+template <typename StorageType>
+void MakeContiguous(std::shared_ptr<StorageType> storage) {}
+
+template <>
+void MakeContiguous(std::shared_ptr<TensorVector<CPUBackend>> storage) {
+  storage->SetContiguous(true);
+}
+
+template <typename Backend>
+struct Backend2Types {};
+
+template <>
+struct Backend2Types<CPUBackend> {
+  using InBackend = CPUBackend;
+  using OutBackend = CPUBackend;
+  using WSInputType = TensorVector<CPUBackend>;
+  using WSOutputType = TensorVector<CPUBackend>;
+};
+
+template <>
+struct Backend2Types<GPUBackend> {
+  using InBackend = GPUBackend;
+  using OutBackend = GPUBackend;
+  using WSInputType = TensorList<GPUBackend>;
+  using WSOutputType = TensorList<GPUBackend>;
+};
+
+template <>
+struct Backend2Types<MixedBackend> {
+  using InBackend = CPUBackend;
+  using OutBackend = GPUBackend;
+  using WSInputType = TensorVector<CPUBackend>;
+  using WSOutputType = TensorList<GPUBackend>;
+};
```

Comment:
Don't we have these somewhere?

Reply:
I couldn't find anything that would do this exactly. Maybe you are thinking about mappings from `dali/pipeline/util/backend2workspace_map.h` or `dali/pipeline/util/op_type_to_workspace.h` or I just missed something.

Reply:
Not exactly in this way. We have some of the information in workspace (templated by the backend), but I guess we are sticking to more strict flow of where the input and output data is placed. In this case I'd say it's fine.